File: processor.js
const fileSystem = require('fs');
const path = require('path');

class UniqueIntProcessor {
    constructor() {
        this.hasAppeared = new Array(2047).fill(false); // Boolean array for integers from -1023 to 1023
        this.minimumInteger = -1023;
    }

    processFile(inputFilePath, outputFilePath) {
        // Reset seen array for each file
        this.hasAppeared = new Array(2047).fill(false);
        const uniqueNumbers = this.readUniqueNumbers(inputFilePath);
        this.writeUniqueNumbers(uniqueNumbers, outputFilePath);
    }

    readUniqueNumbers(inputFilePath) {
        const uniqueNumbers = [];
        const data = fileSystem.readFileSync(inputFilePath, 'utf-8');
        const lines = data.split('\n');
        
        lines.forEach(line => {
            const strippedLine = line.trim();
            if (strippedLine) {
                if (this.isValidIntegerLine(strippedLine)) {
                    const number = parseInt(strippedLine, 10);
                    if (number >= -1023 && number <= 1023) { // Ensure the number is within range
                        if (!this.hasAppeared[number - this.minimumInteger]) {
                            this.hasAppeared[number - this.minimumInteger] = true;
                            uniqueNumbers.push(number);
                        }
                    } else {
                        console.log(`Number out of range: ${number}`);
                    }
                }
            }
        });

        return this.sortUniqueNumbers(uniqueNumbers);
    }

    isValidIntegerLine(line) {
        const number = parseInt(line, 10);
        if (!isNaN(number)) {
            return true;
        } else {
            console.log(`Invalid integer line: ${line}`);
            return false;
        }
    }

    sortUniqueNumbers(numbers) {
        if (numbers.length === 0) {
            return numbers;
        }

        // Implementing Bubble Sort for simplicity
        const count = numbers.length;
        for (let firstIndex = 0; firstIndex < count; firstIndex++) {
            for (let secondIndex = 0; secondIndex < count - firstIndex - 1; secondIndex++) {
                if (numbers[secondIndex] > numbers[secondIndex + 1]) {
                    [numbers[secondIndex], numbers[secondIndex + 1]] = [numbers[secondIndex + 1], numbers[secondIndex]];
                }
            }
        }
        return numbers;
    }

    writeUniqueNumbers(uniqueNumbers, outputFilePath) {
        const data = uniqueNumbers.join('\n') + '\n';
        fileSystem.writeFileSync(outputFilePath, data, 'utf-8');
    }
}

const inputFolder = './Inputs';
const outputFolder = './Output';

const processor = new UniqueIntProcessor();

fileSystem.readdir(inputFolder, (error, files) => {
    if (error) {
        console.error(`Error reading input folder: ${error}`);
        return;
    }

    files.forEach(fileName => {
        if (fileName.endsWith('.txt')) {
            const inputPath = path.join(inputFolder, fileName);
            const outputPath = path.join(outputFolder, `${fileName}_results.txt`);

            // Timing for each file
            const startTime = Date.now();
            processor.processFile(inputPath, outputPath);
            const endTime = Date.now();

            console.log(`Processed ${fileName} in ${(endTime - startTime) / 1000} seconds`);
        }
    });
});

File: processor.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <algorithm>

class UniqueInt {
private:
    std::vector<bool> seen;
    int minInt;
public:
    UniqueInt() : seen(2047, false), minInt(-1023) {}

    void processFile(std::string inputFilePath, std::string outputFilePath) {
        seen.assign(2047, false); // Reset seen array for each file
        std::vector<int> uniqueNumbers = readUniqueIntegers(inputFilePath);
        writeUniqueIntegers(uniqueNumbers, outputFilePath);
    }

private:
    std::vector<int> readUniqueIntegers(std::string inputFilePath) {
        std::vector<int> uniqueNumbers;
        std::ifstream inputFile(inputFilePath);
        if (!inputFile.is_open()) {
            std::cerr << "Error: Unable to open input file: " << inputFilePath << std::endl;
            return uniqueNumbers;
        }
        std::string line;
        while (std::getline(inputFile, line)) {
            std::istringstream iss(line);
            std::string strippedLine;
            iss >> strippedLine;
            if (!strippedLine.empty() && isValidIntegerLine(strippedLine)) {
                int number;
                if (std::stringstream(strippedLine) >> number) {
                    if (-1023 <= number && number <= 1023) { // Ensure the number is within range
                        if (!seen[number - minInt]) {
                            seen[number - minInt] = true;
                            uniqueNumbers.push_back(number);
                        }
                    } else {
                        std::cerr << "Number out of range: " << number << std::endl;
                    }
                }
            }
        }
        inputFile.close();
        std::sort(uniqueNumbers.begin(), uniqueNumbers.end());
        return uniqueNumbers;
    }

    bool isValidIntegerLine(const std::string& line) {
        try {
            std::stoi(line);
            return true;
        } catch (const std::invalid_argument& e) {
            std::cerr << "Invalid integer line: " << line << std::endl;
            return false;
        }
    }

    void writeUniqueIntegers(const std::vector<int>& uniqueNumbers, const std::string& outputFilePath) {
        std::ofstream outputFile(outputFilePath);
        if (!outputFile.is_open()) {
            std::cerr << "Error: Unable to open output file: " << outputFilePath << std::endl;
            return;
        }
        for (int number : uniqueNumbers) {
            outputFile << number << std::endl;
        }
        outputFile.close();
    }
};

int main() {
    UniqueInt uniqueIntProcessor;
    std::string inputFolder = "/root/UniqueInt/Inputs";
    std::string outputFolder = "/root/UniqueInt/Output";

    // Process each file in the input folder
    for (const auto& entry : std::filesystem::directory_iterator(inputFolder)) {
        if (entry.path().extension() == ".txt") {
            std::string inputPath = entry.path().string();
            std::string outputPath = outputFolder + "/" + entry.path().stem().string() + "_results.txt";

            // Timing for each file
            auto startTime = std::chrono::steady_clock::now();
            uniqueIntProcessor.processFile(inputPath, outputPath);
            auto endTime = std::chrono::steady_clock::now();
            std::chrono::duration<double> elapsedTime = endTime - startTime;
            std::cout << "Processed " << inputPath << " in " << elapsedTime.count() << " seconds" << std::endl;
        }
    }

    return 0;
}
